gc/base/MemoryPoolAddressOrderedList.cpp

```diff
@@ -1842,13 +1924,36 @@ MM_MemoryPoolAddressOrderedList::doFreeEntryAlignmentUpTo(MM_EnvironmentBase *en
 		_darkMatterBytes += lostToAlignment;
 	}
 
-	/* update _firstUnalignedFreeEntry and _prevFirstUnalignedFreeEntry */
-	_firstUnalignedFreeEntry = (NULL == currentFreeEntry) ? FREE_ENTRY_END : currentFreeEntry;
-	if (FREE_ENTRY_END != _firstUnalignedFreeEntry) {
-		_prevFirstUnalignedFreeEntry = (NULL == previousFreeEntry) ? FREE_ENTRY_END : previousFreeEntry;
+	/* update _firstCardUnalignedFreeEntry and _prevCardUnalignedFreeEntry */
+	_firstCardUnalignedFreeEntry = (NULL == currentFreeEntry) ? FREE_ENTRY_END : currentFreeEntry;
+	if (FREE_ENTRY_END != _firstCardUnalignedFreeEntry) {
+		_prevCardUnalignedFreeEntry = (NULL == previousFreeEntry) ? FREE_ENTRY_END : previousFreeEntry;
 	} else {
-		_prevFirstUnalignedFreeEntry = FREE_ENTRY_END;
+		_prevCardUnalignedFreeEntry = FREE_ENTRY_END;
 	}
 
 	return alignedLastFreeEntry;
 }
+
+void
+MM_MemoryPoolAddressOrderedList::setParallelGCAlignment(MM_EnvironmentBase *env, bool satbEnabled)
+{
+	if (satbEnabled) {
+		_parallelGCAlignmentBase = _memorySubSpace->getFirstRegion()->getLowAddress();
+		_parallelGCAlignmentSize = _extensions->parSweepChunkSize;
+	} else {
+		_parallelGCAlignmentBase = NULL;
+		_parallelGCAlignmentSize = 0;
+	}
+}
+
+void
+MM_MemoryPoolAddressOrderedList::setSubSpace(MM_MemorySubSpace *memorySubSpace)
+{
+	if (MEMORY_TYPE_OLD == memorySubSpace->getTypeFlags()) {
+		J9HookInterface** mmPrivateHooks = J9_HOOK_INTERFACE(_extensions->privateHookInterface);
+		(*mmPrivateHooks)->J9HookRegisterWithCallSite(mmPrivateHooks, J9HOOK_MM_PRIVATE_CONCURRENT_SATB_TOGGLED, concurrentSATBToggled, OMR_GET_CALLSITE(), (void*)this);
```

Comment:
Just wanted to point out this change I made now... We can't register the hook in initialize method, we have to wait until subspace is set since its not passed through the ctor.

Reply:
I want to confirm that we don't have a case where multiple subspaces are associated with the MP

Reply:
Because this base class method suggests so

```C++
/**
 * Set the owning subspace.
 * This call cascades the owning subspace down through the memorypool hierarchy.
 */
void
MM_MemoryPool::setSubSpace(MM_MemorySubSpace *memorySubSpace)
{
	if(NULL != _children) {
		_children->setSubSpace(memorySubSpace);
	}
	
	if(NULL != _next) {
		_next->setSubSpace(memorySubSpace);
	}
	
	_memorySubSpace = memorySubSpace;
}
```

Reply:
I suppose it fine even if there are multiple subspaces because we filter based on type to regisiter the hook. The issue can be  with unregistering the hook in tearDown, for that we look at the latest subspace set and if it isn't the type OLD, then we miss to unregister.

Reply:
OK  I see, we are not be able to get memory type during pool initialization.
Unfortunate complication, but I have no better suggestion.
Yes, there is indeed only one SubSpace owner/associated with it. MP::getSubSpace() returns just one SubSpace.

Reply:
There might be an alternative approach, to move the registration and the hook in higher level entities: MM_ConfigurationFlat and MM_ConfigurationGenerational. They know exactly when they create the OLD subspace/pool.

They also already have #include on MP, so the callback would be able to call MP::setParallelGCAlignment

This approach would completely free MP from the SATB knowledge (as I said multiple times, I prefer to contain SATB knowledge within higher level entities).

I'm fine with the current approach, but could be moved in a followup change.

Reply:
I'll make a comment about this